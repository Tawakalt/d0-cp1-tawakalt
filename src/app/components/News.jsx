import React from 'react';
import Request from 'superagent';
import _ from 'lodash';
import moment from 'moment';
import UrlStore from '../stores/UrlStore';
import * as UrlActions from '../actions/UrlActions';
import * as AuthActions from '../actions/AuthActions';


export default class News extends React.Component {
  constructor() {
    super();
    this.state = {};
  }

  componentWillMount() {
    UrlStore.on('change', () => {
      this.search();
    });
    this.search();
    this.sources();
  }

  updateSearch() {
    // Fire off action createUrl
    UrlActions.createUrl(this.query.value, this.query2.value);
  }

  updateSearch2() {
    // Fire off action createUrl
    UrlActions.createUrl('test');
  }

  logout() {
    const auth2 = gapi.auth2.getAuthInstance();
    auth2.signOut().then(() => {
      // Fire off action getAuth
      AuthActions.getAuth(null);
    });
  }

  render() {
    const news = _.map(this.state.news, (newss) => {
      // create key
      const id = newss.publishedAt + newss.title;
      // Display the result from the API
      return (
        <div className="container" key={id}>
          <h3 id="l2">{newss.title}</h3>
          <img alt="" src={newss.urlToImage} />
          <p>
            <i>{(moment(new Date(newss.publishedAt))).format('LLLL')}</i>
          </p>
          <p>{newss.author} : {newss.description}</p>
          <p>
            <button className="btn btn-danger">
              <a id="rm" href={newss.url} target="_blank" rel="noopener noreferrer">Read More</a>
            </button>
          </p>
          <hr />
        </div>
      );
    });
    const sources = _.map(this.state.sources, sourcess =>
      // Display the result from the API
       (
         <option key={sourcess.id} value={sourcess.id}>{sourcess.name}</option>

      ));
    return (
      <div>
        <div className="container-fluid" id="news">
          <h1>NEWS!!!</h1>
        </div>
        <div className="container">
          <div className="form-group">
            <div className="row">
              <div className="col-md-4 pull-right">
                <br />
                <button
                  className="btn btn-block btn-danger btn-lg btn-huge"
                  onClick={(event) => { this.logout(); }}
                >
                  Logout
                </button>
              </div>
              <div className="col-md-4 pull-left">
                <label htmlFor="this.query">Sources</label>
                <select
                  className="form-control"
                  ref={(c) => { this.query = c; }}
                  onChange={(event) => { this.updateSearch(); }}
                >
                  {sources}
                </select>
              </div>
              <div className="col-md-4">
                <label htmlFor="this.query2">Sort By</label>
                <select
                  className="form-control"
                  ref={(c) => { this.query2 = c; }}
                  onChange={(event) => { this.updateSearch(); }}
                >
                  <option value="top">Top</option>
                  <option value="latest">Latest</option>
                  <option value="popular">Popular</option>
                </select>
              </div>
            </div>
          </div>
          <div className="row">
            <ul>{news}</ul>
          </div>
        </div>
      </div>
    );
  }

  search() {
    // get url from store
    const url = UrlStore.getUrl();
    Request.get(url).then((response) => {
      this.setState({
        news: response.body.articles,
      });
    });
  }

  sources() {
    // get url from store
    const url = UrlStore.getSourceUrl();
    Request.get(url).then((response) => {
      this.setState({
        sources: response.body.sources,
      });
    });
  }
}
