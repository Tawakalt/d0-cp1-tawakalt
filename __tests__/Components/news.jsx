import React from 'react';
import sinon from 'sinon';
import Request from 'superagent';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import News from '../../src/app/components/News.jsx';
import Utils from '../../src/app/utils';

describe('News', () => {

  const wrapper = shallow(<News />);
  const search = wrapper.instance().search();
  const sources = wrapper.instance().sources();
  const getSort = wrapper.instance().getSort();
  let MockRequest;

  function createRefMock(element) {
  if (element.type === 'select') {
    return {
      focus() {},
    };
  }
  return null;
}

  beforeEach(() => {
    MockRequest = sinon.stub(Request, 'then').callsFake(() => Promise.resolve({ response: 'Successfull' }));
  });
  afterEach(() => {
    MockRequest.restore();
  });

  test('renders correctly', () => {
  const options = {createRefMock};
  const tree = renderer.create(<News />, options);
  expect(tree).toMatchSnapshot();
});

  test('calls method sources', () => {
    Utils.sources();
    wrapper.setState({sources: {}, sourceSortBy: 'top'});
    expect(sources).toEqual(undefined);
  });

  test('calls method search', () => {
    Utils.search();
    wrapper.setState({news: []});
    expect(search).toEqual(undefined);
  });

  test('onSelect sources, function getSort should be called', () => {
    wrapper.find('Select').simulate('select');
    expect(getSort).toEqual(undefined);
  });

  
});
