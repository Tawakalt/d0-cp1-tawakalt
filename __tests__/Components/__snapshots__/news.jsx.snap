// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News Component select input should have a ref attribute 1`] = `
<div>
  <h1 />
</div>
`;
